lib/graphql_metrics/instrumentation.rb

```diff
@@ -1,109 +1,37 @@
-    end
-
-    def setup_instrumentation(schema_definition)
-      schema_definition.instrument(:query, self)
-      schema_definition.instrument(:field, self)
-    end
-
-    def extractor
-      @extractor ||= Extractor.new(self)
-    end
+  module Instrumentation
+    module_function
 
+    # TODO: can this be an instance?
     def before_query(query)
-      return unless extractor_defines_any_visitors?
+      return if query.context[GraphQLMetrics::SKIP_GRAPHQL_METRICS_ANALYSIS]
 
       ns = query.context.namespace(CONTEXT_NAMESPACE)
-      ns[TIMING_CACHE_KEY] = {}
-      ns[START_TIME_KEY] = self.class.current_time
-    rescue StandardError => ex
-      extractor.handle_extraction_exception(ex)
+      ns[GraphQLMetrics::TIMINGS_CAPTURE_ENABLED] = true
+      ns[GraphQLMetrics::INLINE_FIELD_TIMINGS] = {}
+      ns[GraphQLMetrics::LAZY_FIELD_TIMINGS] = {}
     end
 
     def after_query(query)
-      @query = query
-
-      return unless extractor_defines_any_visitors?
-      return if respond_to?(:skip_extraction?) && skip_extraction?(query)
-      return unless @ctx_namespace = query.context.namespace(CONTEXT_NAMESPACE)
-
-      before_query_extracted(query, query.context) if respond_to?(:before_query_extracted)
+      return if query.context[GraphQLMetrics::SKIP_GRAPHQL_METRICS_ANALYSIS]
 
-      extractor.extract!(query)
-
-      after_query_teardown(query) if respond_to?(:after_query_teardown)
-    rescue StandardError => ex
-      extractor.handle_extraction_exception(ex)
-    ensure
-      @query = nil
-    end
-
-    def instrument(type, field)
-      return field unless respond_to?(:field_extracted) || extractor.respond_to?(:field_extracted)
-      return field if type.introspection?
-
-      old_resolve_proc = field.resolve_proc
-      new_resolve_proc = ->(obj, args, ctx) do
-        start_time = self.class.current_time
-        result = old_resolve_proc.call(obj, args, ctx)
-
-        begin
-          next result if respond_to?(:skip_field_resolution_timing?) &&
-            skip_field_resolution_timing?(query, ctx)
-
-          end_time = self.class.current_time
-
-          ns = ctx.namespace(CONTEXT_NAMESPACE)
-
-          ns[TIMING_CACHE_KEY][ctx.ast_node] ||= []
-          ns[TIMING_CACHE_KEY][ctx.ast_node] << end_time - start_time
-
-          result
-        rescue StandardError => ex
-          extractor.handle_extraction_exception(ex)
-          result
-        end
-      end
-
-      field.redefine { resolve(new_resolve_proc) }
-    end
-
-    def after_query_resolver_times(ast_node)
-      ctx_namespace.dig(Instrumentation::TIMING_CACHE_KEY).fetch(ast_node, [])
-    end
-
-    def after_query_start_and_end_time
-      start_time = ctx_namespace[Instrumentation::START_TIME_KEY]
-      return unless start_time
-
-      [start_time, self.class.current_time]
+      ns = query.context.namespace(CONTEXT_NAMESPACE)
+      query_duration = GraphQLMetrics.current_time_monotonic - ns[GraphQLMetrics::QUERY_START_TIME_MONOTONIC]
```

Comment:
I would expect to see `QUERY_START_TIME_MONOTONIC` set in `before_query` for symmetry, but it's handled in the tracer for some reason?

Reply:
From the top of `analyzer.rb`

```ruby
# * Tracer.setup_tracing_before_lexing
# * Tracer.capture_parsing_time
# * Instrumentation.before_query (context setup)
```

I do this in `setup_tracing_before_lexing` because `before_query` doesn't actually run first, and tracing is our first opportunity to truly start these clocks at the beginning of graphql-ruby's phases of execution.